File: example2/src/main/java/com/example2/membership/MembershipRequestTranslator.java
package com.example2.membership;

import org.springframework.stereotype.Component;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

@Component
class MembershipRequestTranslator {
    Membership translate(CreateMembershipRequest createMembershipRequest) {
        if (createMembershipRequest.getBirthDate() == null) {
            throw new RuntimeException("birthDate cannot be empty");
        }

        if (createMembershipRequest.getEmail() == null) {
            throw new RuntimeException("email cannot be empty");
        }

        if (createMembershipRequest.getMemberName() == null) {
            throw new RuntimeException("memberName cannot be empty");
        }

        if (createMembershipRequest.getMembershipEndDate() == null) {
            throw new RuntimeException("membershipEndDate cannot be empty");
        }

        if (createMembershipRequest.getMembershipStartDate() == null) {
            throw new RuntimeException("membershipStartDate cannot be empty");
        }

        return new Membership(
                createMembershipRequest.getMemberName(),
                stringToDate(createMembershipRequest.getBirthDate()),
                createMembershipRequest.getEmail(),
                new DateRange(
                        stringToDate(createMembershipRequest.getMembershipStartDate()).getTime(),
                        stringToDate(createMembershipRequest.getMembershipEndDate()).getTime()
                )
        );
    }

    private Date stringToDate(String stringDate) {
        DateFormat df = new SimpleDateFormat("MM/dd/yyyy");
        try {
            return df.parse(stringDate);
        } catch (ParseException e) {
            throw new RuntimeException("Invalid date format");
        }
    }
}

File: example3/src/main/java/com/example3/membership/MembershipRequestTranslator.kt
package com.example3.membership

import org.springframework.stereotype.Component

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date

@Component
open class MembershipRequestTranslator {

    fun translate(inObject: CreateMembershipRequest) = Membership(
            memberName = inObject.memberName ?: throw RuntimeException("memberName cannot be empty"),
            birthDate = inObject.birthDate?.toDate() ?: throw RuntimeException("birthDate cannot be empty"),
            email = inObject.email ?: throw RuntimeException("email cannot be empty"),
            activeDateRange = inObject.getActiveDateRange()
    )

    private fun CreateMembershipRequest.getActiveDateRange() = DateRange(
            membershipStartDate?.toDate()?.time ?: throw RuntimeException("membershipStartDate cannot be empty"),
            membershipEndDate?.toDate()?.time ?: throw RuntimeException("membershipEndDate cannot be empty")
    )

    private fun String?.toDate(): Date = try {
        SimpleDateFormat("MM/dd/yyyy").parse(this)
    } catch (e: ParseException) {
        throw RuntimeException("Invalid date format")
    }
}
